### PR TITLE
Add Read request in interaction model

### LINKED ISSUE #4430: Add Read request in Interaction Model
#### Problem
Need Read request in Interaction Model to request events or attributes

 #### Proposed Solution
Add Read request implementation.

### PR DESCRIPTION
**Pull request by yunhanw-google**
Problem:
Need read request per IM spec
Refactor MeessageDef

Summary of Changes:
Add read request encoded and unit tests
Split MessageDef into multiple files without logic change for better readability and
maintability

Fixes #4430 #3644

### CODE REVIEW

**Review by rwalker-apple — APPROVED** [review]

File: src/app/MessageDef.cpp [hunk 1]
```diff
@@ -2631,6 +2631,135 @@ AttributeDataList::Builder & AttributeDataList::Builder::EndOfAttributeDataList(
     return *this;
 }
 
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
```
Comment (andy31415) [inline src/app/MessageDef.cpp:2634]: QQ: is this file auto-generated or did you manually add this code?

If this is manually, I have concerns about adding more stuff to a file that got to 3800+ lines of code. Is there something we can do for maintainability?

File: src/app/MessageDef.h [hunk 2]
```diff
@@ -1450,6 +1451,79 @@ class Builder : public ListBuilder
 };
 }; // namespace AttributeDataList
 
+namespace AttributeDataVersionList {
```
Comment (andy31415) [inline src/app/MessageDef.h:1454]: This looks like a stand-alone class. Lets move it to a standalone header (and likely .cpp file as well)

File: src/app/MessageDef.h [hunk 3]
```diff
@@ -1865,6 +1939,136 @@ class Builder : public chip::app::Builder
 };
 }; // namespace InvokeCommand
 
+namespace ReadRequest {
```
Comment (andy31415) [inline src/app/MessageDef.h:1942]: same here: for maintainability, please split up code into independent chunks. This way things can get included/built/unittested independently.

**Review by andy31415 — CHANGES_REQUESTED** [review]

File: src/app/MessageDef.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef.cpp:2634]: manually, i will handle the refactor here in next PR, it should happen next week.
https://github.com/project-chip/connectedhomeip/issues/3644

File: src/app/MessageDef.h (hunk 2 above)
Comment (yunhanw-google) [reply src/app/MessageDef.h:1454]: yep

File: src/app/MessageDef.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/MessageDef.h:1942]: https://github.com/project-chip/connectedhomeip/issues/3644

File: src/app/MessageDef.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef.cpp:2634]: per offline chat, will do the refactor #3644 in this pr

File: src/app/MessageDef.h (hunk 2 above)
Comment (yunhanw-google) [reply src/app/MessageDef.h:1454]: per offline chat, will do the refactor #3644 in this pr

Comment (yunhanw-google) [reply src/app/MessageDef.h]: Done, thanks

File: src/app/MessageDef.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef.cpp]: Done, thanks

File: src/app/MessageDef.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/MessageDef.h]: Done, thanks

**Review by andy31415 — APPROVED** [review]
This is a big moving of items - I will assume it was mechanical and safe, so LGTM.

**Review by BroderickCarlin — APPROVED** [review]

File: src/app/MessageDef/MessageDefHelper.h [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace chip {
+namespace app {
+
+#if CHIP_DETAIL_LOGGING
+
+namespace {
+// this is used to run in signle thread for IM message debug purpose
+uint32_t gPrettyPrintingDepthLevel = 0;
+char gLineBuffer[256];
```
Comment (mspang) [inline src/app/MessageDef/MessageDefHelper.h:64]: This creates a 256 byte static buffer in every translation unit?

It's also thread unsafe. We should not do this, it should be fixed.

Comment (yunhanw-google) [reply src/app/MessageDef/MessageDefHelper.h:64]: It is used to pretty print debug when checking schema validity, and would be disabled in master

Comment (yunhanw-google) [reply src/app/MessageDef/MessageDefHelper.h:64]: Yep, just missed your message, would handle it in https://github.com/project-chip/connectedhomeip/issues/4554

Thanks for your suggestion!

**Pull request merged**